discovery-provider/src/queries/query_helpers.py

```diff
@@ -768,3 +768,186 @@ def get_users_ids(results):
+            Repost.repost_type == type,
+        )
+        .subquery()
+    )
+
+    # Query saves joined against reposts grouped by id and calculate
+    # a combined count for each
+    subquery = (
+        session.query(
+            Save.save_item_id.label('id'),
+            (func.count(Save.save_item_id) + func.count(reposts_count_subquery.c.repost_item_id))
+                .label('count')
+        )
+        # Join against reposts filtering to matching ids.
+        # Inner-join drops no-match ids.
+        .join(
+            reposts_count_subquery,
+            Save.save_item_id == reposts_count_subquery.c.repost_item_id
+        )
+        .filter(
+            Save.is_current == True,
+            Save.is_delete == False,
+            Save.save_type == type
+        )
+        .group_by(
+            Save.save_item_id
+        )
+        .subquery()
+    )
+    return subquery
+
+def create_followee_playlists_subquery(session, current_user_id):
+    """
+    Creates a subquery that returns playlists created by users that
+    `current_user_id` follows.
+
+    Args:
+        session: SQLAlchemy session.
+        current_user_id: The current user id to query against
+    """
+    # Get active followees
+    followee_user_ids_subquery = (
+        session.query(Follow.followee_user_id)
+        .filter(
+            Follow.follower_user_id == current_user_id,
+            Follow.is_current == True,
+            Follow.is_delete == False
+        )
+        .subquery()
+    )
+    followee_playlists_subquery = (
+        session.query(
+            Playlist
+        )
+        .select_from(Playlist)
+        .join(
+            followee_user_ids_subquery,
+            Playlist.playlist_owner_id == followee_user_ids_subquery.c.followee_user_id
+        )
+        .subquery()
+    )
+    return followee_playlists_subquery
+
+
+def seconds_ago(timestamp):
+    """Gets the number of seconds ago `timestamp` was from now as a SqlAlchemy expression."""
+    return func.extract('epoch', (func.now() - timestamp))
+
+def decayed_score(score, created_at, peak = 5, nominal_timestamp = 14 * 24 * 60 * 60):
+    """
+    Creates a decaying (over time) version of the provided `score`. The returned
+    value is score * a multiplier determined by `peak` and `nominal_timestamp`.
+
+    Args:
+        score: (number) The base score to modify
+        created_at: (timestamp) The timestamp the score is attributed to
+        peak?: (number) The peak multipler possible
+
+    Returns:
+        A SQLAlchemy expression representing decayed score (score * multipler)
+        where multipler is represented by:
+        max(0.2, 5 ^ 1 - min(time_ago / nominal_timestamp, 1))
+    """
+    return score * func.greatest(
+        func.pow(
+            5,
+            1 - func.least(seconds_ago(created_at) / nominal_timestamp, 1)
+        ),
+        0.2
+    )
+    
+
+def filter_to_playlist_mood(session, mood, query, correlation = Playlist):
```

Comment:
i don't understand why the model is passed as the default correlation?

Reply:
It's because a default of Playlist is equivalent to no correlation in this case. I thought it made it a little easier to read the query below, but I guess not. Removed the default and pulled it out into a separate query with the correlation optionally appended.